### ISSUE TITLE
INSERT JSON does not check for partition key emptiness

### ISSUE THREAD

**Issue by FortTell** [scylladb/scylla#9853]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [X] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 4.4.4
Cluster size: 5 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian

*Problem:*
1. Empty string is not allowed as a partition key
2. If an entry is inserted into a table via `insert json` mechanism, driver does not check for partition key emptiness and inserts entry with empty partition key

*Reproduce steps:*
Done with Scylla driver `4.13.0.0`
```
create table t (
  a text,
  b text,
  primary key (a)
);

insert into t (a, b) values ('', '1'); -- as expected, fails with `com.datastax.driver.core.exceptions.InvalidQueryException: Key may not be empty`
insert into t json '{"b":""}'; -- as expected, fails with `com.datastax.driver.core.exceptions.InvalidQueryException: Missing mandatory PRIMARY KEY part a`
insert into t json '{"a":null, "b":""}'; -- as expected, fails with `com.datastax.driver.core.exceptions.InvalidQueryException: Missing mandatory PRIMARY KEY part a`
insert into t json '{"a":"", "b":""}'; -- surprisingly, succeeds
select * from t; -- shows 1 entry `a='',b=''`
```

**Comment by slivne** [comment]
@nyh I think you had some patches around NULL in partition key / indexing - I also found this https://github.com/scylladb/scylla/issues/7852 that seems relevant - can you please comment

**Comment by nyh** [comment]
@slivne while this issue is distantly related to #9364, the solution is not related . #9364 is about empty-string partition-key support in materialized views and indexes: There we have an empty-string value for a **regular** (non-key) column, which is allowed, but in a materialized-view or index is "turned" into an empty-string partition key. In contrast, in _the present_ issue, the report is about trying to create a partition with an empty-string partition key, which is officially not allowed.

Issues #7852 is also distantly related to this one but different - #7852 was about trying to insert a **null* (missing) partition key. This is not the same as an empty-string partition key, but indeed the end-result in both cases should be the same - an error messsage.

We have tests in `test_null.py`: `test_insert_empty_string_key`, `test_update_empty_string_key`, `test_delete_empty_string_key`, which verify that trying to operate on an empty-string partition key with `INSERT`, `UPDATE` and `DELETE` is not allowed, and indeed it isn't. But we don't have the same test for `INSERT JSON` and we need to create such a test, which will fail (according to the above report), and then we need to make it pass. I'll do that now.

**Comment by nyh** [comment]
A simple test reproducing this issue (passing on Cassandra, failing on Scylla):

```python
def test_insert_json_empty_string_key(cql, table1):
    s = random_string()
    with pytest.raises(InvalidRequest, match='Key may not be empty'):
        cql.execute("""INSERT INTO %s JSON '{"p": "", "c": "%s", "v": "cat"}'""" % (table1, s))
```

**Comment by nyh** [comment]
Sent a patch to the mailing list titled `cql: INSERT JSON should refuse empty-string partition key`.
The patch is one line of code, with many times more than that for tests and explanations :-)

**Comment by FortTell** [comment]
Thank you for fixing this issue. Which driver version will include the fix?

**Comment by nyh** [comment]
The fix is not in the driver, but in the Scylla server itself. It will be in the nightly build, and later in release 4.7 and probably backported to earlier branches (stay tuned here on information on when it's backported).

**Comment by nyh** [comment]
We have to backport this patch to every relevant branch... It was reported by one user, and we just got a report a report from another user who hit the same bug. And it wasn't just an annoyance that empty keys got inserted - it also caused `sstable::set_first_and_last_keys() ` to throw "first key of summary is empty" and fail to write sstables.

**Comment by nyh** [comment]
The fix (commit 8fd5041092dce452e536d58eff710aae24f5af97) is already in branch-5.0.
Backported now to next-4.6 (2f2d22a8641a029ba34cb030e189085c820f4ee8) and next-4.5 (865cfebfed19d38317ae806e6354eba5ad2fefe1).
